### PR TITLE
[Uniques V2] Tips

### PR DESCRIPTION
**Pull request by jsidorenko**
This PR adds a new `pay_tips` method.
Since the royalties shouldn't be enforced on the pallet/contract level, we still need a way to allow marketplaces to set them optionally. One possible solution would be to use batched transactions, but then it won't be possible for any UI/BCX to distinguish between tips paid and regular payments. This PR solves this by adding a new `pay_tips` method and emitting the `TipSent` event.

### CODE REVIEW

**Comment by gilescope** [comment]
"This PR adds a new param to the buy_item method." - it does not currently do that, it adds a new method to the pallet. Tbh I was expecting an Option<Tip> on the buy_item method. Does the tip go to the NFT creator automatically or does it go to the marketplace?

https://github.com/paritytech/substrate/blob/5a48536a49859f83dbbbab1285311f14c192e7b1/frame/support/src/traits/tokens/misc.rs#L142 - we seem to have bitfields to indicate why we are sending money - perhaps a ROYALTY flag could be added there? But then there's no reference to point out which NFT it's a tip for.

**Comment by jsidorenko** [comment]
Yeah, if you check the commits history, initially I put an `Option<Tips>` param into the `item_buy`. Then I realised there could be more methods those would want to get the optional tips added like `buy_from_seller` or any other method which is not there yet. By extracting the tips payment into a separate method we could allow to combine that easily with any other method without modifying the params.

**Comment by jsidorenko** [comment]
> But then there's no reference to point out which NFT it's a tip for.
I think this is critical for a lot of external systems to know the NFT's id when the tip is sent

**Comment by jsidorenko** [comment]
I've converted tuple into struct as per Oliver suggestion

**Comment by jsidorenko** [comment]
/cmd queue -c bench-bot $ pallet dev pallet_nfts

**Comment by ruseinov** [comment]
Looks good to me

File: frame/nfts/src/benchmarking.rs [hunk 1]
```diff
@@ -458,5 +458,28 @@ benchmarks_instance_pallet! {
...
+	pay_tips {
+		let n in 0 .. T::MaxTips::get() as u32;
+		let amount = BalanceOf::<T, I>::from(100u32);
+		let caller: T::AccountId = whitelisted_caller();
+		let collection = T::Helper::collection(0);
+		let item = T::Helper::item(0);
+		let tips: BoundedVec<_, _> = vec![
+			ItemTip
+				{ collection, item, receiver: caller.clone(), amount }; T::MaxTips::get() as usize
+		].try_into().unwrap();
```
Comment (ggwpez) [inline frame/nfts/src/benchmarking.rs:470]:
```suggestion
		let tips: BoundedVec<_, _> = bounded_vec![
			ItemTip
				{ collection, item, receiver: caller.clone(), amount }; n as usize
		];
```

**Review by ggwpez — APPROVED** [review]

**Review by ruseinov — APPROVED** [review]

File: frame/nfts/src/benchmarking.rs (hunk 1 above)
Comment (jsidorenko) [reply frame/nfts/src/benchmarking.rs]: It looks like we can't use bounded_vec! macro in benchmarks, probably because of `#[cfg(feature = "std")]`

**Comment by jsidorenko** [comment]
/cmd queue -c bench-bot $ pallet dev pallet_nfts

**Comment by jsidorenko** [comment]
bot merge

**Pull request merged**